src/checkers/inference/solver/constraintgraph/ConstraintGraph.java

```diff
@@ -0,0 +1,112 @@
+package checkers.inference.solver.constraintgraph;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import checkers.inference.model.Constraint;
+import checkers.inference.model.Slot;
+import checkers.inference.model.SubtypeConstraint;
+import checkers.inference.model.VariableSlot;
+
+/**
+ * ConstraintGraph represents constraints in a graph form. Each constraint is an
+ * edge, and each slot is a vertex. ConstraintGraph is used for separating
+ * constraint into different components by running graph traversal algorithm on
+ * it. The graph is undirected, but for subtype constraint we use special edge
```

Comment:
In general the graph should be a directed graph, with most constraints being bi-directional edges.

At minimum, we should clean up the code comments to reflect this.